### PR TITLE
HDDS-6764: EC: DN ability to create RECOVERING containers for EC reconstruction.

### PR DESCRIPTION
**Pull request by umamaheswararao**
## What changes were proposed in this pull request?

This PR proposes to add the createREcoveringContainer API in ContainerProtocolCalls, which can be used to store reconstruction in progress containers.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6764

## How was this patch tested?

Added tests to verify.

### CODE REVIEW

File: hadoop-hdds/interface-client/src/main/proto/DatanodeClientProtocol.proto [hunk 1]
```diff
@@ -270,6 +271,8 @@ message  CreateContainerRequestProto {
   repeated KeyValue metadata = 2;
   optional ContainerType containerType = 3 [default = KeyValueContainer];
   optional int32 replicaIndex = 4;
+  //optional LifeCycleState state = 5;
```
Comment (adoroszlai) [inline hadoop-hdds/interface-client/src/main/proto/DatanodeClientProtocol.proto:274]: Do we need to add this?

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/ozone/container/ContainerTestHelper.java [hunk 2]
```diff
@@ -436,6 +441,7 @@ public static Builder newPutBlockRequestBuilder(Pipeline pipeline,
 
     ContainerProtos.PutBlockRequestProto.Builder putRequest =
         ContainerProtos.PutBlockRequestProto.newBuilder();
+    putRequest.setEof(true);
```
Comment (adoroszlai) [inline hadoop-hdds/common/src/test/java/org/apache/hadoop/ozone/container/ContainerTestHelper.java:444]: Why is `eof=true` needed here?  I think caller can set it later.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/HddsDispatcher.java [hunk 3]
```diff
@@ -471,7 +473,9 @@ public void validateContainerCommand(
     }
 
     State containerState = container.getContainerState();
-    if (!HddsUtils.isReadOnly(msg) && containerState != State.OPEN) {
+    if (!HddsUtils.isReadOnly(msg)
+        && (containerState != State.OPEN
+        && containerState != State.RECOVERING)) {
```
Comment (adoroszlai) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/HddsDispatcher.java:478]: Would it make sense to add a method for handling OPEN and RECOVERING state the same way in specific cases?  (I'm not sure what a good name for the method might be, though.)

**Review by adoroszlai — COMMENTED** [review]
Thanks @umamaheswararao for the patch.

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/ozone/container/ContainerTestHelper.java (hunk 2 above)
Comment (umamaheswararao) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/ozone/container/ContainerTestHelper.java]: Ah, I think I added when debugging.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/HddsDispatcher.java (hunk 3 above)
Comment (umamaheswararao) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/HddsDispatcher.java]: I tried to export into a method. Let's see if that's make sense.

File: hadoop-hdds/interface-client/src/main/proto/DatanodeClientProtocol.proto (hunk 1 above)
Comment (umamaheswararao) [reply hadoop-hdds/interface-client/src/main/proto/DatanodeClientProtocol.proto]: removed it. Below state is there.

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/ozone/container/ContainerTestHelper.java (hunk 2 above)
Comment (umamaheswararao) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/ozone/container/ContainerTestHelper.java]: Ah, I think I added this while testing.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/HddsDispatcher.java (hunk 3 above)
Comment (umamaheswararao) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/HddsDispatcher.java]: Tried to export into a method. Let's see if that make sense.

**Comment by umamaheswararao** [comment]
Thanks @adoroszlai for the review. Please check the latest updates.

**Review by adoroszlai — APPROVED** [review]
Thanks @umamaheswararao for updating the patch, LGTM.

**Comment by adoroszlai** [comment]
@umamaheswararao it looks like `TestOzoneECClient` is failing consistently in CI.

https://github.com/apache/ozone/runs/6702568728#step:5:1644
https://github.com/apache/ozone/runs/6705930506#step:5:1654
https://github.com/apache/ozone/runs/6709532641#step:5:1654
https://github.com/apache/ozone/runs/6710406479#step:5:1654
https://github.com/apache/ozone/runs/6711102895#step:5:1654

**Comment by umamaheswararao** [comment]
That tests got added recently. I don't think it's related, but there is some flakeyness in the tests. Tests were added from this HDDS-6808

cc: @kaijchen 
Should we revert that tests until we figure out the cause?

**Comment by umamaheswararao** [comment]
I have just rebased with that change. Let's see. If this reappears, then I will add @Ignore tag.

**Comment by umamaheswararao** [comment]
Thanks @adoroszlai for the reviews!

**Pull request merged**